pypesto/ensemble/ensemble.py

```diff
@@ -747,3 +749,97 @@ def check_identifiability(self) -> pd.DataFrame:
+                      max_size: int,
+                      max_per_start: int,):
+    """
+    Creates the indices of each start that will be included
+    in the ensemble.
+
+    Parameters
+    ----------
+    fval_trace:
+        the fval-trace of each start.
+    cutoff:
+        Exclude parameters from the optimization if the nllh
+        is higher than the `cutoff`.
+    max_size:
+        The maximum size the ensemble should be.
+    max_per_start:
+        The maximum number of vectors to be included from a
+        single optimization start.
+
+    Returns
+    -------
+        A list of number of candidates per start that are to
+        be included in the ensemble.
+
+    """
+    # choose possible candidates
+    ens_ind = [np.flatnonzero(fval <= cutoff) for fval in fval_trace]
+
+    # count the number of candidates per start
+    n_per_starts = np.array([len(start) for start in ens_ind])
+
+    # if all possible indices can be included, return
+    if (n_per_starts < max_per_start).all() and sum(n_per_starts) < max_size:
+        return ens_ind
+
+    # trimm down starts that exceed the limit:
+    n_per_starts = [min(n, max_per_start) for n in n_per_starts]
+
+    # trimm down more until it fits the max size
+    decr = 0
+    while(sum(n_per_starts) > max_size):
+        n_per_starts = [min(n, max_per_start-decr)
+                        for n in n_per_starts]
+        decr += 1
+    # TODO: Possibility. With this implementation we could
+    #  in a scenario, where we have more candidates than
+    #  max size end up with an ensemble of size
+    #  `max_size - len(n_starts)` in the worst case. We could introduce
+    #  a flag which would be `force_max`, that indicates
+    #  whether those remaining free slots should be filled by
+    #  entries from certain starts. This would brng up the
+    #  discussion which starts to choose. One obvious choice
+    #  would be the best starts based on their endpoint.
+
+    return n_per_starts
+
+
+def indices_per_start(trace_start: np.ndarray,
+                      cutoff: float,
+                      n_per_start: int,
+                      dist_or_best: bool,):
+    """
+    Returns the indices to be taken into an ensemble.
+
+    Parameters
+    ----------
+    trace_start:
+        The fval_trace of a single start.
+    cutoff:
+        Exclude parameters from the optimization if the nllh
+        is higher than the `cutoff`.
+    n_per_start:
+        The number of indices to be included from one start.
+    dist_or_best:
+        Boolean flag, whether the best (False) values from the
+        start should be taken or whether the indices should be
+        more evenly distributed.
+
+    Returns
+    -------
+        The indices which to include in the ensemble.
+    """
+
+    candidates = np.flatnonzero(trace_start <= cutoff)
+
+    if dist_or_best:
+        indices = np.round(np.linspace(0, len(candidates)-1, n_per_start))
+        return candidates[indices.astype(int)]
```

Comment:
Maybe convert to set? Otherwise is it possible that e.g. `len(candidates)-1==1` and `n_per_start==6` so `indices==[0, 0, 0, 1, 1, 1]` and so the returned vector indices are duplicates?

Reply:
no n_per_start will give you a number that will definitely be lower than len(candidates) as per code so this should not be a problem. I do like lists more but that is a personal preference :D

Reply:
Ah, if it's not intended for a user to call this method directly, then the name can be prefixed with a `_` e.g. `_indices_per_start`